db/src/excision.rs

```diff
@@ -0,0 +1,129 @@
+    AEVTrie,
+    filter_aev_to_eav,
+};
+
+use schema::{
+    SchemaBuilding,
+};
+
+use types::{
+    PartitionMap,
+};
+
+/// Details about an excision:
+/// - a target to excise (for now, an entid);
+/// - a possibly empty set of attributes to excise (the empty set means all attributes, not no
+///   attributes);
+/// - and a possibly omitted transaction ID to limit the excision before.  (TODO: check whether
+///   Datomic excises the last retraction before the first remaining assertion, and make our
+///   behaviour agree.)
+///
+/// `:db/before` doesn't make sense globally, since in Mentat, monotonically increasing
+/// transaction IDs don't guarantee monotonically increasing txInstant values.  Therefore, we
+/// accept only `:db/beforeT` and allow consumers to turn `:db/before` timestamps into
+/// transaction IDs in whatever way they see fit.
+#[derive(Clone, Debug, Default)]
+pub(crate) struct Excision {
+    pub(crate) target: Entid,
+    pub(crate) attrs: Option<BTreeSet<Entid>>,
+    pub(crate) before_tx: Option<Entid>,
+}
+
+pub(crate) type Excisions = BTreeMap<Entid, Excision>;
+
+/// Extract excisions from the given transacted datoms.
+pub(crate) fn excisions<'schema>(partition_map: &'schema PartitionMap, schema: &'schema Schema, aev_trie: &AEVTrie<'schema>) -> Result<Option<Excisions>> {
+    let pair = |a: Entid| -> Result<(Entid, &'schema Attribute)> {
+        schema.require_attribute_for_entid(a).map(|attribute| (a, attribute))
+    };
+
+    if aev_trie.contains_key(&pair(entids::DB_EXCISE_BEFORE)?) {
+        bail!(DbError::BadExcision(":db.excise/before".into())); // TODO: more details.
+    }
+
+    let eav_trie = filter_aev_to_eav(aev_trie, |&(a, _)|
+                                     a == entids::DB_EXCISE ||
+                                     a == entids::DB_EXCISE_ATTRS ||
+                                     a == entids::DB_EXCISE_BEFORE_T);
+
+    let mut excisions = BTreeMap::default();
+
+    for (&e, avs) in eav_trie.iter() {
+        for (&(_a, _attribute), ars) in avs {
+            if !ars.retract.is_empty() {
+                bail!(DbError::BadExcision("retraction".into())); // TODO: more details.
+            }
+        }
+
+        let target = avs.get(&pair(entids::DB_EXCISE)?)
+            .and_then(|ars| ars.add.iter().next().cloned())
+            .and_then(|v| v.into_entid())
+            .ok_or_else(|| DbError::BadExcision("no :db/excise".into()))?; // TODO: more details.
+
+        if schema.get_ident(target).is_some() {
+            bail!(DbError::BadExcision("cannot mutate schema".into())); // TODO: more details.
+        }
+
+        let partition = partition_map.partition_for_entid(target)
+            .ok_or_else(|| DbError::BadExcision("target has no partition".into()))?; // TODO: more details.
+        // Right now, Mentat only supports `:db.part/{db,user,tx}`, and tests hack in `:db.part/fake`.
+        if partition == ":db.part/db" || partition == ":db.part/tx" {
```

Comment:
Explicitly testing for partitions that are allowed in excisions seems more future-proof.

Reply:
You're right.  I'll grow an `excision_allowed()` flag.

Reply:
I've updated to use this.